### ISSUE TITLE
Enum where all variants have the same variable(s)

### ISSUE THREAD

**Issue by Pyriphlegethon** [Manishearth/rust-clippy#368]
```rust
enum SomeEnum {
    A(i32),
    B(i32),
    C(i32)
}
```
should really be 
```rust
enum SomeEnum {
    A,
    B,
    C
}

struct SomeStruct {
    variant: SomeEnum,
    data: i32
}
```
in most cases.

**Comment by birkenfeld** [comment]
I'm not sure. I would say this is a matter of how you prefer to conceptualise the data.

**Comment by llogiq** [comment]
Yeah. As a basic guard against false positives, we should check that there is some getter function that returns the data out of all variants. A stronger check would be that all accesses to the contained values are done regardless of variant (i.e. all code that uses the `data` either calls or copies the getter). If we can do that, I'd suggest we can make the lint `Warn`, otherwise I vote for `Allow`.

**Comment by Pyriphlegethon** [comment]
Well, we could also make this lint complain only if the enum has more than 4 (or any other number, really) variants.

**Comment by Manishearth** [comment]
Not sure the situation changes at _any_ number of variants.

**Comment by llogiq** [comment]
Hmmm, that could work, and be much easier than checking access patterns. We'd need to test it on code in the wild. So let's make it `Allow` first, check rustc, servo and a few other code bases with it, and then set it to `Warn` unless we find false positives.

**Comment by frewsxcv** [comment]
I personally think this is a subjective code structuring/organization preference more so than a lint that informs me of bad practices. There are both pros and cons to each and I could see both being used in certain scenarios.
